kulkunen/drivers/securitas.py

```diff
@@ -0,0 +1,173 @@
+import jsonschema
+import requests
+from datetime import timedelta
+from django.core.exceptions import ValidationError
+from django.utils import timezone
+from jsonschema.exceptions import ValidationError as JsonSchemaValidationError
+
+from .base import AccessControlDriver
+
+
+class SecuritasDriver(AccessControlDriver):
+    BASE_URL = "https://extapi.flow.securitas.com/v1/"
+
+    SYSTEM_CONFIG_SCHEMA = {
+        "type": "object",
+        "properties": {
+            "api_key": {
+                "type": "string",
+            },
+        },
+    }
+
+    RESOURCE_CONFIG_SCHEMA = {
+        "type": "object",
+        "properties": {
+            "uses_pincode": {
+                "type": "boolean",
+                "default": True,
+            },
+        },
+    }
+
+    DEFAULT_RESOURCE_CONFIG = {
+        "uses_pincode": True,
+    }
+
+    def install_grant(self, grant):
+        self.logger.info("[%s] Installing Securitas grant", grant)
+
+        assert grant.state == grant.INSTALLING
+
+        user = grant.reservation.user
+
+        user_id = str(user.pk)
+
+        access_user, _ = self.system.users.get_or_create(
+            identifier=user_id,
+            user=user,
+        )
+
+        params = {
+            "resourceId": grant.resource.identifier,
+            "validFrom": grant.starts_at.isoformat(),
+            "validTo": grant.ends_at.isoformat(),
+        }
+
+        response = self._api_post(
+            "access",
+            {
+                "userExtId": user_id,
+                "firstName": user.first_name,
+                "lastName": user.last_name,
+                "email": user.email,
+                "notifyUser": True,
+                **params,
+            },
+        )
+        grant.identifier = response.json()["accessId"]
+
+        # now try and create pincode
+        if self.get_resource_setting(grant.resource, "uses_pincode"):
+            response = self._api_post("codes", params)
+            data = response.json()
+            grant.access_code = data["code"]
+            # save the Securitas ID to driver data so we can delete later
+            grant.driver_data = {"code_id": data["id"]}
```

Comment:
"code_id" is the primary key on Securitas' side, distinct from the actual pincode.